kep/42-podgroup-coscheduling/README.md

```diff
@@ -0,0 +1,125 @@
+ 
+## Use Cases
+Batch workloads such as Spark jobs, TensorFlow jobs that have to run altogether.
+
+## Design Details
+
+### PodGroup
+
+We define a CRD name PodGroup to help schedule, its definition is as follows:
+```build
+// PodGroupSpec represents the template of a pod group.
+type PodGroupSpec struct {
+	// MinMember defines the minimal number of members/tasks to run the pod group;
+	// if there's not enough resources to start all tasks, the scheduler
+	// will not start anyone.
+	MinMember uint32 `json:"minMember"`
+
+	// MaxScheduleTime defines the maximal time of members/tasks to wait before run the pod group;
+	MaxScheduleTime *metav1.Duration `json:"maxScheduleTime,omitempty"`
+}
+
+// PodGroupStatus represents the current state of a pod group.
+type PodGroupStatus struct {
+	// Current phase of PodGroup.
+	Phase PodGroupPhase `json:"phase"`
+
+	// OccupiedBy marks the podgroup occupied by which group.
+	// Owner reference would be used to filled it, if not initialize, it is empty
+	OccupiedBy string `json:"occupiedBy,omitempty"`
+
+	// The number of actively running pods.
+	// +optional
+	Scheduled uint32 `json:"scheduled"`
+
+	// The number of actively running pods.
+	// +optional
+	Running uint32 `json:"running"`
+
+	// The number of pods which reached phase Succeeded.
+	// +optional
+	Succeeded uint32 `json:"succeeded"`
+
+	// The number of pods which reached phase Failed.
+	// +optional
+	Failed uint32 `json:"failed"`
+
+	// ScheduleStartTime of the group
+	ScheduleStartTime metav1.Time `json:"scheduleStartTime"`
+}
+```
+
+### Controller
+
+We define a controller to reconcile PodGroup status, and we can query the job status through describing the PodGroup. Onece a pod in a group failed, the Group Status is marked Failed. Controller would also help recover from abnormal cases, e.g. batch scheduling is interrupted due to
+cluster upgrade.
+
+### Extension points
+
+#### QueueSort
+
+To make sure a group of pods can be scheduled as soon as possible. We implemented this `extension point`. The main progress is as follows:
+
+1. Sort based on Pod Priority
+2. Pod does not belong to any group
+3. PodGroup creation time
+4. Pod creation time
+
+#### PreFilter
+
+This extension helps pre-filter pods. It is useful, especially when there are not enough resources in a cluster. The overall flow works as below:
+
+1. Allow the pods that do not belong to any group.
+2. If there are no groups scheduling, we check resource, if enough, we allow the pod.
+3. If there are groups running, we check if the current pod belong the group having the max progress(num(Pods)/minMember), if it is, we allow it.
+4. Otherwise, we check if the max finished group can still run when allow this pod. If we can, allow it.
+5. Otherwise, we check if the pod has higher priority compared with the max finished one. If yes, we reject the pod belongs to the group and allow the current one. 
```

Comment:
```suggestion
3. If there are other pod groups not fully scheduled, we check if the pod group current pod belongs to is the closest to its completion - i.e., we check its completion progress by `len(waiting pods of a pod group) / pod group's minMember`.
    - if it is, this pod is allowed.
    - if it's not, we check if allowing it would still make the top-progressed pod group (or you want to say all progressed pod grouped?) scheduled:
        - if so, this pod is allowed
        - if not, we check if the current pod has higher priority than the top-progressed pod group (I guess for convenience, you would still need the `PriorityClass` defined in PodGroup?). If it does, this pod is allowed and the top-progressed pod group gets rejected.
```

And remove bullet 4 & 5.

Reply:
I don't agree to use `len(waiting pods of a pod group) / pod group's minMember` to check if the pg is the closest to its completion. The minMember and the resource requested in pod are different. Like this：
PG1: min 10 waiting 5 --> 50%
PG2 min 100 waiting 60 --> 60%

Reply:
@cwdsuzhou does the rewording comment https://github.com/kubernetes-sigs/scheduler-plugins/pull/42/files#r481413128 make sense to you?

Reply:
Regarding @denkensk 's concern, I'd leave it for @cwdsuzhou to comment. Probably it's not a bad idea to also look at the absolute number to reach completion.

Reply:
Before pods owned by other groups start scheduling, we would check whether the group with the max completion can continue scheduling when we schedule these pods. According `Sort` func. Current method may be more reasonable. Otherwise, the final scheduling results may conflict with the `Sort`.

Reply:
I think it's a good idea to find the podGroup most likely to be successfully scheduled . But I think the current method is not reasonable. Because the different pods may request different resource and with different scale. And also with different priority.
I think we need to think about the pod.resource.requested、absolute number and the priority to find the the podGroup most likely to be successfully scheduled.
@cwdsuzhou @Huang-Wei  WDYT?